python/torch_mlir/extras/fx_importer.py

```diff
@@ -208,34 +209,58 @@
 }
 
 
-def sparsity_encoding(
-    shape: torch.Size, sparsity: tuple[torch.layout, int, int]
-) -> str:
-    """Returns sparse tensor encoding for the given sparse layout as string.
+@dataclass(frozen=True)
+class Sparsity:
+    """Class for keeping track of sparsity meta data."""
 
-    The method currently just supports 2-dim sparse formats. This should be
-    generalized to the torch.sparse encodings for prefix dense batch dimensions
-    and suffix dense subtensor dimensions. Since MLIR supports a superset of what
-    is currently implememented in torch.sparse, this should not a be problem.
-    """
+    layout: torch.layout
+    batch_dim: int
+    sparse_dim: int
+    dense_dim: int
+    pos_width: int
+    crd_width: int
+
+
+def sparsity_encoding(shape: torch.Size, sparsity: Sparsity) -> str:
+    """Returns sparse tensor encoding for the given sparse layout as string."""
     assert sparsity is not None
-    sparse_layout, posw, crdw = sparsity
-
-    # TODO: any rank
-    if len(shape) != 2:
-        raise RuntimeError(f"Unsupported sparse rank {len(shape)}")
-
-    if sparse_layout is torch.sparse_coo:
-        smap = f"(i,j)->(i:compressed(nonunique),j:singleton)"
-    elif sparse_layout is torch.sparse_csr:
-        smap = f"(i,j)->(i:dense,j:compressed)"
-    elif sparse_layout is torch.sparse_csc:
-        smap = f"(i,j)->(j:dense,i:compressed)"
+
+    # Sparse tensors have the form
+    #   [ <batch_dimensions> , <sparse_dimensions>, <dense_dimensions> ]
+    # which map directly to MLIR types.
+    batch_dim, sparse_dim, dense_dim = (
+        sparsity.batch_dim,
+        sparsity.sparse_dim,
+        sparsity.dense_dim,
+    )
+    dim = batch_dim + sparse_dim + dense_dim
+    assert dim == len(shape)
+
+    dims = ",".join(f"d{d}" for d in range(0, dim))
+
+    if sparsity.layout is torch.sparse_coo:
+        assert sparse_dim == 2  # TODO: deeper sparse dims
+        lvls = f"d{batch_dim}:compressed(nonunique),d{batch_dim+1}:singleton"
+    elif sparsity.layout is torch.sparse_csr:
+        assert sparse_dim == 2
+        lvls = f"d{batch_dim}:dense,d{batch_dim+1}:compressed"
+    elif sparsity.layout is torch.sparse_csc:
+        assert sparse_dim == 2
+        lvls = f"d{batch_dim+1}:dense,d{batch_dim}:compressed"
     else:
         # TODO: block format (derive block size!)
         raise RuntimeError(f"Unsupported sparse layout {sparse_layout}")
 
-    return f"#sparse_tensor.encoding<{{map={smap},posWidth={posw},crdWidth={crdw}}}>"
+    if batch_dim > 0:
+        batch = ",".join(f"d{d}:dense" for d in range(0, batch_dim))
+        lvls = f"{batch},{lvls}"
+
+    if dense_dim > 0:
```

Comment:
nit: Are these condition necessary? `for d in range(0, 0)` seems to work even when batch_dim == 0 as well.

Reply:
yeah, except for adding the , in between batch and lvls in that case

Reply:
(but I find that other people are usually much better in writing very concise python, so let me know if I missed a cleverl trick)